[ENHANCEMENT] Support for keycloak versions that no longer have /auth in the context path

Newer versions of Keycloak based on Quarkus distribution no longer use /auth as the first path after the hostname.

e.g. currently the helm chart generates based on these values:
global:
  identity:
    keycloak:
      url:
        protocol: "http"
        host: "keycloak"
        port: "8081"

The following env var (just one of the various places /auth shows up)
          - name: IDENTITY_AUTH_PROVIDER_BACKEND_URL
            value: "http://keycloak:8081/auth/realms/camunda-platform"

But in our use case it should be
            value: "http://keycloak:8081/realms/camunda-platform"

For all URLs based on publicIssuerUrl it is already working good since this setting specifies the full url including path to the realm.

I have attached the values file I have been testing with, where I overrode the various env vars regenerated with /auth in them to have url's without /auth. This allows Identity to now connect to the external keycloak and it seems to have created the realm camunda-platform with the user demo, which also has been granted the roles for Tasklist, Optimize, Operate and Identity.

[values-external-keycloak.txt](https://github.com/camunda/camunda-platform-helm/files/9879015/values-external-keycloak.txt)

After this I can login to identity on https://identity.external.com which redirects to keycloak, login with demo/demo and I get redirected to identity.

For https://operate.external.com from this example the same, how ever I end up with no permission. If I check the operate logs it shows:

2022-10-27 11:41:14.833 ERROR 8 --- [nio-8080-exec-7] i.c.o.w.s.i.IdentityController           : Error in authentication callback: 
io.camunda.identity.sdk.authentication.exception.InvalidClaimException: the provided claims are invalid
...
Caused by: com.auth0.jwt.exceptions.IncorrectClaimException: The Token can't be used before 2022-10-27T11:41:16Z.

I guess of of the cluster nodes clock is out of sync, but other than that everything seems to work with keycloak 19.

Update: The clock was indeed 2 seconds out of sync. Fixed that and operate login works fine now.


**Describe the enhancement/feature:**

Please support making the /auth bit of the path optional (or even configurable in case people have their keycloak instances under a custom path?)

**Desired outcome and acceptance tests:**

A way to configure an external keycloak that has no /auth in the URL.

--- a/charts/camunda-platform/README.md
+++ b/charts/camunda-platform/README.md
@@ -96,7 +96,8 @@ Check out the default [values.yaml](values.yaml) file, which contains the same c
 | | `identity.keycloak.url.protocol` |    Can be used to set existing Keycloak URL protocol | `` |
 | | `identity.keycloak.url.host` |    Can be used to set existing Keycloak URL host | `` |
 | | `identity.keycloak.url.port` |    Can be used to set existing Keycloak URL port | `` |
-| | `identity.keycloak.auth` |   Can be used incorporate with "global.identity.keycloak.url" and "identity.keycloak.enabled: false" set existing Keycloak URL instead of the one comes with Camunda Platform Helm chart | `{}"` |
+| | `identity.keycloak.contextPath` |   Defines the endpoint of Keycloak which varies between Keycloak versions. In Keycloak v16.x.x it's hard-coded as '/auth', but in v19.x.x it's '/' | `"/auth"` |
+| | `identity.keycloak.auth` |   Can be used incorporate with "global.identity.keycloak.url" and "identity.keycloak.enabled: false" set existing Keycloak URL instead of the one comes with Camunda Platform Helm chart | `{}` |
 | | `identity.keycloak.auth.adminUser` |    Can be used to configure admin user to access existing Keycloak | `""` |
 | | `identity.keycloak.auth.existingSecret` |    Can be used to configure admin user password by using existing secret with key `admin-password` to access existing Keycloak | `""` |
 | `elasticsearch`| `enabled` | Enable Elasticsearch deployment as part of the Camunda Platform Cluster | `true` |

--- a/charts/camunda-platform/charts/identity/templates/_helpers.tpl
+++ b/charts/camunda-platform/charts/identity/templates/_helpers.tpl
@@ -157,15 +157,23 @@ This is mainly used for cases where the external Keycloak is used.
 {{- end -}}
 
 {{/*
+[identity] Get Keycloak contextPath based on global value.
+*/}}
+{{- define "identity.keycloak.contextPath" -}}
+    {{ .Values.global.identity.keycloak.contextPath | default "/auth" }}
+{{- end -}}
+
+{{/*
 [identity] Get Keycloak full URL (protocol, host, and port).
 */}}
 {{- define "identity.keycloak.url" -}}
     {{- include "identity.keycloak.isConfigured" . -}}
     {{-
-      printf "%s://%s:%s"
+      printf "%s://%s:%s%s"
         (include "identity.keycloak.protocol" .)
         (include "identity.keycloak.service" .)
         (include "identity.keycloak.port" .)
+        (include "identity.keycloak.contextPath" .)
     -}}
 {{- end -}}
 

--- a/charts/camunda-platform/charts/identity/templates/deployment.yaml
+++ b/charts/camunda-platform/charts/identity/templates/deployment.yaml
@@ -112,7 +112,7 @@ spec:
           - name: SERVER_PORT
             value: "8080"
           - name: KEYCLOAK_URL
-            value: "{{ include "identity.keycloak.url" . }}/auth"
+            value: {{ include "identity.keycloak.url" . | quote }}
           - name: IDENTITY_AUTH_PROVIDER_ISSUER_URL
             value: {{ .Values.global.identity.auth.publicIssuerUrl | quote }}
           - name: IDENTITY_AUTH_PROVIDER_BACKEND_URL

--- a/charts/camunda-platform/templates/_helpers.tpl
+++ b/charts/camunda-platform/templates/_helpers.tpl
@@ -75,10 +75,10 @@ Subcharts can't access values from other sub-charts or the parent, global only. 
 */}}
 
 {{- define "camundaPlatform.issuerBackendUrl" -}}
-    {{- $keycloakRealmPath := "/auth/realms/camunda-platform" -}}
+    {{- $keycloakRealmPath := "/realms/camunda-platform" -}}
     {{- if .Values.global.identity.keycloak.url -}}
         {{- include "identity.keycloak.url" . -}}{{- $keycloakRealmPath -}}
     {{- else -}}
-        http://{{ include "common.names.dependency.fullname" (dict "chartName" "keycloak" "chartValues" . "context" $) | trunc 20 | trimSuffix "-" }}:80{{ $keycloakRealmPath }}
+        http://{{ include "common.names.dependency.fullname" (dict "chartName" "keycloak" "chartValues" . "context" $) | trunc 20 | trimSuffix "-" }}:80{{- include "identity.keycloak.contextPath" . -}}{{ $keycloakRealmPath }}
     {{- end -}}
 {{- end -}}

--- a/charts/camunda-platform/templates/ingress.yaml
+++ b/charts/camunda-platform/templates/ingress.yaml
@@ -24,8 +24,7 @@ spec:
                 name: {{ include "identity.keycloak.service" .Subcharts.identity }}
                 port:
                   number: {{ include "identity.keycloak.port" .Subcharts.identity }}
-            # It's complex to change the context path in Keycloak v16.x.x, so it's hard-coded.
-            path: /auth
+            path: {{ include "identity.keycloak.contextPath" .Subcharts.identity }}
             pathType: Prefix
           - backend:
               service:

--- a/charts/camunda-platform/test/ingress_test.go
+++ b/charts/camunda-platform/test/ingress_test.go
@@ -12,6 +12,7 @@ import (
 	"github.com/stretchr/testify/suite"
 	v1 "k8s.io/api/apps/v1"
 	v12 "k8s.io/api/core/v1"
+	netv1 "k8s.io/api/networking/v1"
 )
 
 type ingressTemplateTest struct {
@@ -36,7 +37,7 @@ func TestIngressTemplate(t *testing.T) {
 	})
 }
 
-func (s *ingressTemplateTest) TestIngressEnabledAndKeycloakChartHasNecessarySettings() {
+func (s *ingressTemplateTest) TestIngressEnabledAndKeycloakChartProxyForwardingEnabled() {
 	// given
 	options := &helm.Options{
 		SetValues: map[string]string{
@@ -64,6 +65,46 @@ func (s *ingressTemplateTest) TestIngressEnabledAndKeycloakChartHasNecessarySett
 		v12.EnvVar{
 			Name:  "KEYCLOAK_PROXY_ADDRESS_FORWARDING",
 			Value: "true",
+		})
+}
+
+func (s *ingressTemplateTest) TestIngressEnabledWithKeycloakCustomContextPath() {
+	// given
+	options := &helm.Options{
+		SetValues: map[string]string{
+			"global.ingress.enabled":               "true",
+			"global.identity.keycloak.contextPath": "/custom",
+			"identity.contextPath":                 "/identity",
+			"identity.keycloak.enabled":            "true",
+		},
+		KubectlOptions: k8s.NewKubectlOptions("", "", s.namespace),
+		ExtraArgs:      map[string][]string{"template": {"--debug"}, "install": {"--debug"}},
+	}
+
+	// when
+	ingressOutput := helm.RenderTemplate(s.T(), options, s.chartPath, s.release, s.templates)
+
+	var ingress netv1.Ingress
+	helm.UnmarshalK8SYaml(s.T(), ingressOutput, &ingress)
+
+	// then
+	path := ingress.Spec.Rules[0].HTTP.Paths[0]
+	s.Require().Equal("/custom", path.Path)
+	s.Require().Equal("camunda-platform-tes", path.Backend.Service.Name)
+
+	// when
+	extraArgs := []string{"--show-only", "charts/identity/charts/keycloak/templates/statefulset.yaml"}
+	stsOutput := helm.RenderTemplate(s.T(), options, s.chartPath, s.release, nil, extraArgs...)
+
+	var statefulSet v1.StatefulSet
+	helm.UnmarshalK8SYaml(s.T(), stsOutput, &statefulSet)
+
+	// then
+	env := statefulSet.Spec.Template.Spec.Containers[0].Env
+	s.Require().Contains(env,
+		v12.EnvVar{
+			Name:  "KEYCLOAK_HTTP_RELATIVE_PATH",
+			Value: "/custom",
 		})
 }
 

--- a/charts/camunda-platform/values.yaml
+++ b/charts/camunda-platform/values.yaml
@@ -82,14 +82,17 @@ global:
   # Identity configuration to configure identity specifics on global level, which can be accessed by other sub-charts
   identity:
     keycloak:
-      # Identity.keycloak.url is can be used incorporate with "identity.keycloak.enabled: false" to use your own Keycloak instead of the one comes with Camunda Platform Helm chart.
+      # Identity.keycloak.url can be used incorporate with "identity.keycloak.enabled: false" to use your own Keycloak instead of the one comes with Camunda Platform Helm chart.
       url: {}
         # Example to produce the following URL "https://keycloak.prod.svc.cluster.local:8443":
         # url:
         #   protocol: "https"
         #   host: "keycloak.prod.svc.cluster.local"
         #   port: "8443"
-        # Identity.keycloak.auth same as "identity.keycloak.auth" but it's used for existing Keycloak.
+      # Identity.keycloak.contextPath defines the endpoint of Keycloak which varies between Keycloak versions.
+      # In Keycloak v16.x.x it's hard-coded as '/auth', but in v19.x.x it's '/'.
+      contextPath: "/auth"
+      # Identity.keycloak.auth same as "identity.keycloak.auth" but it's used for existing Keycloak.
       auth: {}
         # Identity.keycloak.auth.adminUser can be used to configure admin user to access existing Keycloak.
         # adminUser: ""
@@ -849,10 +852,14 @@ identity:
   keycloak:
     # Keycloak.enabled is used incorporate with "global.identity.keycloak" to use your own Keycloak instead of the one comes with Camunda Platform Helm chart.
     enabled: true
-    # KEYCLOAK_PROXY_ADDRESS_FORWARDING is set to "true" to make it work with Ingress (if it's enabled).
     extraEnvVars:
+    # KEYCLOAK_PROXY_ADDRESS_FORWARDING is set to "true" to make it work with Ingress (if it's enabled).
     - name: KEYCLOAK_PROXY_ADDRESS_FORWARDING
       value: "true"
+    # KEYCLOAK_HTTP_RELATIVE_PATH is valid for v19.x.x only and it's added for compatibility between Keycloak versions
+    # where in Keycloak v16.x.x it's hard-coded as '/auth', but in v19.x.x it's '/'.
+    - name: KEYCLOAK_HTTP_RELATIVE_PATH
+      value: "{{ .Values.global.identity.keycloak.contextPath }}"
     # Keycloak.ingress configuration, to setup a separated Ingress for Keycloak if needed.
     # It's disabled by default and just added here for visibility.
     ingress:

